### ISSUE TITLE
4.4.0 regression / ?flush does not respect SS_BASE_URL

### ISSUE THREAD

**Issue by dnsl48** [silverstripe/silverstripe-framework#9061]
## Affected Version

Silverstripe Version: 4.4.0 and 4.4.1
Wamp: 3.1.13
Apache: 2.4.33
PHP: 7.2.4
MySQL: 5.7.21
MariaDB: 10.2.14

## Description

Original description: https://forum.silverstripe.org/t/4-4-1-redirects-to-documentroot-instead-of-ss-base-url/1986

Question: Why do I get redirected to DocumentRoot instead of SS_BASE_URL, unlike previous versions?

After a successful install of SilverStripe 4.4.1, I get auto-redirected to DocumentRoot instead of the project’s base url. I can navigate back to the correct sub-directory and everything functions as normal.

This issue does not occur in versions 4.2.0, 4.3.0, and 4.3.3

My DocumentRoot is C:\wamp\www reachable at 127.0.0.1
Version 3.4.4 is in C:\wamp\www\344\ and is reachable at 127.0.0.1/334/
Version 4.4.1 is in C:\wamp\www\441\ and is reachable at 127.0.0.1/441/

They were obtained through composer like so:

```
C:\wamp\www>composer create-project silverstripe/installer 433 4.3.3
C:\wamp\www>composer create-project silverstripe/installer 441 4.4.1
```

My DocumentRoot’s index.php is a welcome page with links that takes you to either version.

The following illustrates each step in the installation process:

![00da23ed36b4377c96bf01b9ce7bda5a2ba440cf](https://user-images.githubusercontent.com/9313746/59472059-4a550a00-8e91-11e9-9e7b-4ff69985f00a.jpeg)
4.3.3 installs successfully and redirects correctly.

![2](https://user-images.githubusercontent.com/9313746/59472062-4d4ffa80-8e91-11e9-9f6d-88992d0aa7af.jpeg)
4.4.1 installs successfully but redirects to DocumentRoot. I manually type in the missing sub-directory to get to the last step.

## Notes

I would also like to note that if I do flush=all while not in dev mode, I get redirected to DocumentRoot and with a urlspecialstoken, but if I manually type in the missing sub-directory and with the provided urlspecialstoken, it will flush successfully.

## PRs

 - [x] https://github.com/silverstripe/silverstripe-framework/pull/9074

**Comment by UndefinedOffset** [comment]
I also observed this without setting the SS_BASE_URL environment setting, and just letting SilverStripe figure it out on it's own.

**Comment by dnsl48** [comment]
> I also observed this without setting the SS_BASE_URL environment setting while flushing site without coming from the installer.

Could you give some more details about how to reproduce?

**Comment by UndefinedOffset** [comment]
I have a site setup say `http://localhost/test-site` it was temporarily in live mode and does not use the `SS_BASE_URL` environment setting in `.env` when you add `?flush=1` to the url it redirects to `http://localhost/dev/confirm` instead of `http://localhost/test-site/dev/confirm`.

**Comment by dnsl48** [comment]
TL;DR;
`SS_BASE_URL` is required for redirects in anyway. It might have only been working for you because flush didn't redirect to the login form (you already been authenticated).

---
To be consistent I will recap a couple things.

 - The default value of `SS_ENVIRONMENT_TYPE` is `live`.
 - SilverStripe 4.4 changed behaviour around `dev/*` urls, `?flush`, `?isTest` and `?isDev`. In `live` mode it now requires users to explicitly confirm their actions on the form placed at `/dev/confirm`.

It also requires users to be authenticated to perform those actions (unless the database is not initialized).

How it worked before (SilverStripe <= 4.3):
In `dev` mode `?flush` happens _in-place_ without redirects.
In `live` mode:
 - framework checks authentication and if user is not logged in, it redirects to `Security/login`
 - If user is authenticated, everything works fine as `?flush` happens _in-place_.

How it works now (SilverStripe >= 4.4):
In `dev` mode `?flush` happens _in-place_ without redirects.
In `live` mode:
  - framework checks authentication and if user is not logged in, it redirects to `Security/login`
  - if user is authenticated, it redirects to `dev/confirm` (the confirmation form)
  - the confirmation form action redirects back to the original page requested with `?flush` (or /dev/*)

In both cases, redirects to the authentication and the confirmation forms won't work without `SS_BASE_URL` defined. It only happened to be working before because you've been authenticated and it didn't redirect for flush.

**Comment by aNickzz** [comment]
For an site located at `http://localhost/test-site`, without `SS_BASE_URL` set;

Navigating to `http://localhost/test-site?flush` redirects `http://localhost/test-site/Security/login` correctly, but after you login it redirects to `http://localhost/dev/confirm/url-specials` instead of `http://localhost/test-site/dev/confirm/url-specials`.

So the redirect to `url-specials` must be doing something different to the redirect for `login`.

Even after setting `SS_BASE_URL="http://localhost/test-site"` in `.env` the `url-specials` redirect (and the one back to the original page) are not working; they're doing the same as when `SS_BASE_URL` was not set.

Tested in `4.4.1`.

**Comment by ScopeyNZ** [comment]
Hold one, wasn't the fix for this issue in `4.4.2`?

**Comment by ScopeyNZ** [comment]
Here's the commit for reference: https://github.com/silverstripe/silverstripe-framework/commit/7ef13e7ef60a935dc75a72a5b63bb49311d469d4

Please try with 4.4.2 an we can re-open if it's still an issue.

**Comment by dnsl48** [comment]
Yes, for some reasons it didn't make it into `4.4.1`
